persistent-postgresql/Database/Persist/Postgresql.hs

```diff
@@ -1738,6 +1754,176 @@ repsertManySql ent n = putManySql' conflictColumns fields ent n
+  [HandleUpdateCollision record] ->
+  -- | A list of the updates to apply that aren't dependent on the record being inserted.
+  [Update record] ->
+  -- a filter condition
+  [Filter record] ->
+  ReaderT backend m ()
+upsertManyWhere [] _ _ _ = return ()
+upsertManyWhere records fieldValues updates filters = do
+  conn <- asks projectBackend
+  uncurry rawExecute $
+    mkBulkUpsertQuery records conn fieldValues updates filters
+
+-- | This creates the query for 'upsertManyWhere'. If you
+-- provide an empty list of updates to perform, then it will generate
+-- a dummy/no-op update using the first field of the record. This avoids
+-- duplicate key exceptions.
+mkBulkUpsertQuery
+    :: (PersistEntity record, PersistEntityBackend record ~ SqlBackend)
+    => [record] -- ^ A list of the records you want to insert, or update
+    -> SqlBackend
+    -> [HandleUpdateCollision record] -- ^ A list of the fields you want to copy over.
+    -> [Update record] -- ^ A list of the updates to apply that aren't dependent on the record being inserted.
+    -> [Filter record]
+    -> (Text, [PersistValue])
+mkBulkUpsertQuery records conn fieldValues updates filters =
+  (q, recordValues <> updsValues <> copyUnlessValues)
+  where
+    mfieldDef x = case x of
+        CopyField rec -> Right (fieldDbToText (persistFieldDef rec))
+        CopyUnlessEq rec val -> Left (fieldDbToText (persistFieldDef rec), toPersistValue val)
+    (fieldsToMaybeCopy, updateFieldNames) = partitionEithers $ map mfieldDef fieldValues
+    fieldDbToText = escapeF . fieldDB
+    entityDef' = entityDef records
+    firstField = case entityFieldNames of
+        [] -> error "The entity you're trying to insert does not have any fields."
+        (field:_) -> field
+    entityFieldNames = map fieldDbToText (entityFields entityDef')
+    nameOfTable = escapeE . entityDB $ entityDef'
+    copyUnlessValues = map snd fieldsToMaybeCopy
+    recordValues = concatMap (map toPersistValue . toPersistFields) records
+    recordPlaceholders = Util.commaSeparated $ map (Util.parenWrapped . Util.commaSeparated . map (const "?") . toPersistFields) records
+    mkCondFieldSet n _ = T.concat [n, "=EXCLUDED.", n]
+    condFieldSets = map (uncurry mkCondFieldSet) fieldsToMaybeCopy
+    fieldSets = map (\n -> T.concat [n, "=EXCLUDED.", n, ""]) updateFieldNames
+    upds = map (Util.mkPostgresUpdateText (escapeF) id) updates
+    updsValues = map (\(Update _ val _) -> toPersistValue val) updates
+    wher = if null filters then "" else filterClause False conn filters
+    updateText = case fieldSets <> upds <> condFieldSets of
+        [] -> T.concat [firstField, "=EXCLUDED.", firstField]
+        xs -> Util.commaSeparated xs
+    q = T.concat
+        [ "INSERT INTO "
+        , nameOfTable
+        , Util.parenWrapped . Util.commaSeparated $ entityFieldNames
+        , " VALUES "
+        , recordPlaceholders
+        , " ON CONFLICT "
+        , Util.parenWrapped $ firstField
```

Comment:
I _think_ this is wrong, shouldn't the conflict field be conditional to what the user wants to pass in, and not guaranteed to be the first field?

Reply:
So, the way `upsert` works is that it requires a *single* `Unique` key for the record, and that `Unique` is what is used to trigger the `update` part instead of the `insert` part. If the entity has many `Unique` keys, you're required to use `upsertBy`, which accepts the `Unique` in question.

MySQL doesn't explicitly list the conflict target - it triggers if *any* uniqueness key is duplicated. 

I wonder what happens if there's a uniqueness constraint that is violated by an `INSERT` that isnt' listed here. I'd imagine it just errors. Worth investigating!

I think we can just list *all* of the uniqueness keys here - the primary key + all `Unique`s that persistent knows about.

Reply:
donezo